buffer/src/main/java/io/netty/buffer/api/Buffer.java

```diff
@@ -0,0 +1,703 @@
+     *                     this many bytes. This way, the buffer can grow by more than what is immediately necessary,
+     *                     thus amortising the costs of allocating and copying.
+     * @param allowCompaction {@code true} if the method is allowed to modify the
+     *                                   {@linkplain #readerOffset() reader offset} and
+     *                                   {@linkplain #writerOffset() writer offset}, otherwise {@code false}.
+     * @throws BufferReadOnlyException if this buffer is {@linkplain #readOnly() read-only}.
+     * @throws IllegalArgumentException if {@code size} or {@code minimumGrowth} are negative.
+     * @throws IllegalStateException if this buffer is in a bad state.
+     */
+    Buffer ensureWritable(int size, int minimumGrowth, boolean allowCompaction);
+
+    /**
+     * Returns a copy of this buffer's readable bytes.
+     * Modifying the content of the returned buffer will not affect this buffers contents.
+     * The two buffers will maintain separate offsets. This method is identical to
+     * {@code buf.copy(buf.readerOffset(), buf.readableBytes())}.
+     * This method does not modify {@link #readerOffset()} or {@link #writerOffset()} of this buffer.
+     * <p>
+     * The copy is created with a {@linkplain #writerOffset() write offset} equal to the length of the copied data,
+     * so that the entire contents of the copy is ready to be read.
+     *
+     * @return A new buffer instance, with independent {@link #readerOffset()} and {@link #writerOffset()},
+     * that contains a copy of the readable region of this buffer.
+     * @throws BufferClosedException if this buffer is closed.
+     */
+    default Buffer copy() {
+        int offset = readerOffset();
+        int length = readableBytes();
+        return copy(offset, length);
+    }
+
+    /**
+     * Returns a copy of the given region of this buffer.
+     * Modifying the content of the returned buffer will not affect this buffers contents.
+     * The two buffers will  maintain separate offsets.
+     * This method does not modify {@link #readerOffset()} or {@link #writerOffset()} of this buffer.
+     * <p>
+     * The copy is created with a {@linkplain #writerOffset() write offset} equal to the length of the copy,
+     * so that the entire contents of the copy is ready to be read.
+     *
+     * @param offset The offset where copying should start from. This is the offset of the first byte copied.
+     * @param length The number of bytes to copy, and the capacity of the returned buffer.
+     * @return A new buffer instance, with independent {@link #readerOffset()} and {@link #writerOffset()},
+     * that contains a copy of the given region of this buffer.
+     * @throws IllegalArgumentException if the {@code offset} or {@code length} reaches outside the bounds of the
+     * buffer.
+     * @throws BufferClosedException if this buffer is closed.
+     */
```

Comment:
If a read-only Buffer is copied, does the copy end up read only too?

Reply:
Thinking about it, maybe it should not, because then this can be used to get out of the otherwise irreversible read-only state.